Framework/HistogramData/src/Rebin.cpp

```diff
@@ -0,0 +1,154 @@
+      inew++; /* old and new bins do not overlap */
+    else if (xo_high <= xn_low)
+      iold++; /* old and new bins do not overlap */
+    else {
+      // delta is the overlap of the bins on the x axis
+      auto delta = xo_high < xn_high ? xo_high : xn_high;
+      delta -= xo_low > xn_low ? xo_low : xn_low;
+
+      if (delta <= 0.0) {
+        throw std::runtime_error("Negative or zero overlaps not allowed.");
+      }
+      auto factor = 1 / owidth;
+      ynew[inew] += yold[iold] * delta * factor;
+      enew[inew] += eold[iold] * eold[iold] * delta * factor;
+
+      if (xn_high > xo_high) {
+        iold++;
+      } else {
+        inew++;
+      }
+    }
+  }
+
+  return Histogram(binEdges, newCounts,
+                   CountStandardDeviations(std::move(newCountVariances)));
+}
+
+Histogram rebinFrequencies(const Histogram &input, const BinEdges &binEdges) {
+  auto &xold = input.x();
+  auto &yold = input.y();
+  auto &eold = input.e();
+
+  auto &xnew = binEdges.rawData();
+  Frequencies newFrequencies(xnew.size() - 1);
+  FrequencyStandardDeviations newFrequencyStdDev(xnew.size() - 1);
+  auto &ynew = newFrequencies.mutableData();
+  auto &enew = newFrequencyStdDev.mutableData();
+
+  auto size_yold = yold.size();
+  auto size_ynew = ynew.size();
+  size_t iold = 0;
+  size_t inew = 0;
+
+  while ((inew < size_ynew) && (iold < size_yold)) {
+    auto xo_low = xold[iold];
+    auto xo_high = xold[iold + 1];
+    auto xn_low = xnew[inew];
+    auto xn_high = xnew[inew + 1];
+
+    auto owidth = xo_high - xo_low;
+    auto nwidth = xn_high - xn_low;
+
+    if (owidth <= 0.0 || nwidth <= 0.0)
+      throw std::runtime_error("Negative or zero bin widths not allowed.");
+
+    if (xn_high <= xo_low)
+      inew++; /* old and new bins do not overlap */
+    else if (xo_high <= xn_low)
+      iold++; /* old and new bins do not overlap */
+    else {
+      //        delta is the overlap of the bins on the x axis
+      auto delta = xo_high < xn_high ? xo_high : xn_high;
+      delta -= xo_low > xn_low ? xo_low : xn_low;
+
+      if (delta <= 0.0) {
+        throw std::runtime_error("Negative or zero bin overlaps not allowed.");
+      }
+
+      ynew[inew] += yold[iold] * delta;
+      enew[inew] += eold[iold] * eold[iold] * delta * owidth;
+
+      if (xn_high > xo_high) {
+        iold++;
+      } else {
+        auto factor = 1 / nwidth;
+        ynew[inew] *= factor;
+        enew[inew] = sqrt(enew[inew]) * factor;
+        inew++;
+      }
+    }
+  }
+
+  return Histogram(binEdges, newFrequencies, newFrequencyStdDev);
+}
+} // anonymous namespace
+
+namespace Mantid {
+namespace HistogramData {
+
+/** Rebins data according to a new set of bin edges.
+* @param input :: input histogram data to be rebinned.
+* @param binEdges :: input will be rebinned according to this set of bin edges.
+* @returns The rebinned histogram.
+* @throws std::runtime_error if the input histogram xmode is not BinEdges and
+* the input yMode is undefined.
```

Comment:
or if the bin edges in either input or output are yield non-positive bin widths.